### PR TITLE
Change representation of signed integer values to use sign extension

### PR DESCRIPTION
**Pull request by jafingerhut**

### CODE REVIEW

**Comment by jafingerhut** [comment]
Intended as a fix for https://github.com/p4lang/p4runtime/issues/111

**Comment by antoninbas** [comment]
I'm going to be reviewing this before the end of the week. Adding @teverman as well since he wrote the original version of the padding scheme and he is good at spotting corner cases :).

**Comment by jafingerhut** [comment]
No rush on my part, and glad to have more sets of eyes on such changes, in case I've missed something.

File: docs/v1/P4Runtime-Spec.mdk [hunk 1]
```diff
@@ -1565,53 +1565,85 @@ fits the encoded integer value. This representation achieves three goals:
...
+bit to represent, regardless of whether it is signed or unsigned.
+
+The shortest possible binary string for an integer $V$ that needs $A$
+bits to represent it is computed as:
 ~ Begin CPP
-if (value != 0)
-  encoded_string_size = (high_bit + 8) / 8;
-else
-  encoded_string_size = 1;
+encoded_string_size = floor((A + 7) / 8)
```
Comment (teverman) [inline docs/v1/P4Runtime-Spec.mdk]: Would it be clearer to rename encoded_string_size to minimum_string_size?  Just a thought.

File: docs/v1/P4Runtime-Spec.mdk [hunk 2]
```diff
@@ -1565,53 +1565,85 @@ fits the encoded integer value. This representation achieves three goals:
...
+
+Upon receiving a binary string, the P4Runtime receiver (whether the
+server or the client) does not impose any restrictions on the length
+of the string itself. Instead, the receiver verifies that the value
+encoded by the string fits within the expected type (signed or
+unsigned) and P4Info-specified bitwidth for the P4 object value.
+
+For received values expected to fit within a `bit<W>` type, the value
+is in range if, after removing all most significant 0 bits, the value
+remaining is `W` bits or less.
```
Comment (teverman) [inline docs/v1/P4Runtime-Spec.mdk]: I think it's more accurate to specify the value width here, i.e "the remaining value's width is W bits or less".

File: docs/v1/P4Runtime-Spec.mdk [hunk 3]
```diff
@@ -1565,53 +1565,85 @@ fits the encoded integer value. This representation achieves three goals:
...
+
+For received values expected to fit within an `int<W>` type, the value
+is in range if, after "undoing sign extension", the value remaining is
+`W` bits or less. To undo sign extension, start by eliminating the
```
Comment (teverman) [inline docs/v1/P4Runtime-Spec.mdk]: Add width per previous comment.

**Comment by jafingerhut** [comment]
Closing this PR in favor of this newer one.  It has identical contents: https://github.com/p4lang/p4runtime/pull/151

Reason: I deleted the Github fork that had the branch containing this PR, and creating a new PR was the most straightforward way I know to make further changes to it.

**Pull request closed without merging**